src/frontend/src/components/authenticateBox.ts

```diff
@@ -0,0 +1,227 @@
+ * to II or to another dapp */
+export const authenticateBox = (
+  connection: Connection,
+  templates: AuthTemplates
+): Promise<LoginData> => {
+  const retryOnError = async (result: LoginFlowResult): Promise<LoginData> => {
+    switch (result.tag) {
+      case "err":
+        await displayError({
+          title: result.title,
+          message: result.message,
+          detail: result.detail !== "" ? result.detail : undefined,
+          primaryButton: "Try again",
+        });
+        return authenticateBox(connection, templates);
+      case "ok":
+        return result;
+      case "canceled":
+        return authenticateBox(connection, templates);
+      default:
+        unreachable(result);
+        break;
+    }
+  };
+
+  return new Promise<LoginData>((resolve) => {
+    authenticateBoxTemplate({
+      templates,
+      addDevice: () => addRemoteDevice(connection),
+      onContinue: async (userNumber) => {
+        const loginData = await authenticate(connection, userNumber).then(
+          retryOnError
+        );
+        setUserNumber(loginData.userNumber);
+        resolve(loginData);
+      },
+      register: async () => {
+        const loginData = await registerIfAllowed(connection).then(
+          retryOnError
+        );
+        setUserNumber(loginData.userNumber);
+        resolve(loginData);
+      },
+      recoverAnchor: (userNumber) => useRecovery(connection, userNumber),
+      userNumber: getUserNumber(),
+    });
+  });
+};
+
+export const authenticateBoxTemplate = (props: PageProps): void => {
+  const container = document.getElementById("pageContent") as HTMLElement;
+  render(pageContent(props), container);
+};
+
+const pageContent = ({
+  templates,
+  addDevice,
+  onContinue,
+  recoverAnchor,
+  register,
+  userNumber,
+}: PageProps): TemplateResult => {
+  const anchorInput = mkAnchorInput({
+    userNumber,
+    onSubmit: onContinue,
+  });
+
+  return html` <div class="l-container c-card c-card--highlight">
+      <!-- The title is hidden but used for accessibility -->
+      <h1 class="is-hidden">Internet Identity</h1>
+      <div class="c-logo">${icLogo}</div>
+      <div class="l-stack t-centered">
+        ${templates.message} ${templates.chasm ? mkChasm(templates.chasm) : ""}
+      </div>
+
+      ${anchorInput.template}
+
+      <button
+        @click="${anchorInput.submit}"
+        id="authorizeButton"
+        class="c-button"
+      >
+        Authorize
+      </button>
+      ${mkLinks({
+        register,
+        recoverAnchor: () => recoverAnchor(anchorInput.readUserNumber()),
+        addDevice,
+      })}
+    </div>
+    ${footer}`;
+};
+
+export const authenticate = async (
+  connection: Connection,
+  userNumber: bigint
+): Promise<LoginFlowSuccess | LoginFlowError> => {
+  try {
+    const result = await withLoader(() => connection.login(userNumber));
+    return apiResultToLoginFlowResult(result);
+  } catch (error) {
+    return {
```

Comment:
not in this pr but doesn't this comes often in the code base? an utils can maybe be extracted to map error

Reply:
definitely; will clean this up a bit after this is merged now that I have a better idea of the error handling flow